runtime/shared_heap.c

```diff
@@ -481,50 +493,118 @@ static intnat pool_sweep(struct caml_heap_state* local, pool** plist,
 
   {
     header_t* p = POOL_FIRST_BLOCK(a, sz);
-    header_t* end = POOL_END(a);
-    mlsize_t wh = wsize_sizeclass[sz];
+    header_t* last_free_block = NULL;
+    const header_t* end = POOL_END(a);
+    const mlsize_t wh = wsize_sizeclass[sz];
     int all_used = 1;
     struct heap_stats* s = &local->stats;
 
-    /* conceptually, this is incremented by [wh] for every iteration
-       below, however we can hoist these increments knowing that [p ==
-       end] on exit from the loop (as asserted) */
+    a->next_obj = 0;
+
     work = end - p;
     do {
-      header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
-      if (hd == 0) {
-        /* already on freelist */
-        all_used = 0;
-      } else if (Has_status_hd(hd, caml_global_heap_state.GARBAGE)) {
+      header_t hd = (header_t)*p;
+
+      if( p + 4096 < end ) {
+        caml_prefetch(p + 4096);
```

Comment:
`p` is a `header_t *`, so `p + 4096` is 4 Kwords away, i.e. 32 KiB. This seems like a long way ahead to be prefetching (and in fact won't it always be outside the pool?). My impression is that the intention is "one page ahead", so try `p + 512` (or `(char*)p + caml_plat_pagesize`, to make the intention clear)?

Reply:
Uh oh. Thanks. It's surprising there was a performance change despite this. Will fix and re-run.

Reply:
Oops! Fixed in https://github.com/ocaml/ocaml/pull/13616/commits/c26799fb684e7641f22afe32ebba72349a26e897 and am re-running the benchmarks now (also including prefetching in trunk sweeping).